### DISCUSSION TITLE
Custom Api Url?

### DISCUSSION THREAD

**Discussion by Andoramb** [allanbunch/node-red-openai-api#8] [Ideas]
Hi!

Would it be possible to provide custom API url when setting up the node?
This could potentially play along when running local llm models, such as text-generation-webui-docker or LM  Studio.

Then tokens don't really matter.

Thanks :)

**Comment by allanbunch** [comment]
Hey there! Thanks for the suggestion.

I'll look into this. My initial thought is that this would take the project in a new direction--which I am very much open to as it evolves. As things currently exist, the node is quite specific to OpenAI's REST API endpoints, request paths, and payload, so we'd be looking at more than simply providing an option for a custom API URL in the node options.

Be sure to update to the latest version of the node if you haven't done so in the past day. I fixed several initial bugs with the OpenAI Assistants requests.

Cheers!

**Comment by Andoramb** [comment]
Sure, great initiative!👍 
I am experimenting with running LLM locally with [text-generation-webui](https://github.com/oobabooga/text-generation-webui).
And for what I've seen, the API implementation is taken from OpenAI API, a few examples [here](https://github.com/oobabooga/text-generation-webui/wiki/12-%E2%80%90-OpenAI-API#examples)
Payload, requests look all similar 😃 

So at a first try, I just wanted to put these 2 together, by changing the base url.

I probably could do the changes locally for me, but wanted to know your opinion and do a collab 😄 

And as Oobabooga evolves, it might get additional features as well.

**Reply by allanbunch** [reply]
Got it! Thanks for the details. I ran a few tests with the node's domain pointing to a local Text Generation Web UI (TGWU 😀) instance and things are looking good! I'm able to chat with the [zephyr-7B-beta-GGUF](https://huggingface.co/TheBloke/zephyr-7B-beta-GGUF) model in a flow by injecting an OpenAI API _messages_ payload.

I'll send a note your way after I add the option to to the node and publish a preview branch.

**Reply by allanbunch** [reply]
@Andoramb 

Branch is up! You can install it and give it a go using:

```bash
npm install github:allanbunch/node-red-openai-api#feature/custom-service-host
```

You'll find that the node now has a configurable "Service Host". Let me know how things go for you. Thanks!

<img width="377" alt="service host node properties" src="https://github.com/allanbunch/node-red-openai-api/assets/4503640/2a41d60d-4b9e-4db7-bc06-f9c99d927118">


<img width="371" alt="service host node config" src="https://github.com/allanbunch/node-red-openai-api/assets/4503640/2ffcb1fe-f4d1-4500-9c96-31c60b0b0965">

<img width="371" alt="OpenAI service host node properties" src="https://github.com/allanbunch/node-red-openai-api/assets/4503640/66d8f810-127e-43c7-861a-6e2941fab03b">

<img width="373" alt="OpenAI service host node config" src="https://github.com/allanbunch/node-red-openai-api/assets/4503640/4bb76a74-89fa-4dd8-9bd9-5177d4594a4f">

**Comment by Andoramb** [comment]
Awesome! 🎉
I'll check it as soon as I get the chance 🔥

**Reply by allanbunch** [reply]
Excellent! 🚀
I'll mention that you suggested this feature in my next NPM release if you like. Just let me know!

**Comment by Andoramb** [comment]
So let's see here:

1. I did not manage to connect it to text-generation-webui, but it's probably something with my setup:
`"POST /v1/chat/completions HTTP/1.1" 404 -`
(I tried with the example for also from chat completions, tried it with postman but still not found)

2. With LM studio it went better:
![image](https://github.com/allanbunch/node-red-openai-api/assets/3399628/910efb57-5eae-4ed7-99ec-de98a2e12e86)
_Chat completion_ worked as expected, but probably other functions are not implemented yet.
But it works! 😄 🎉

You don't really have to mention me, you are developing it, I just barely sit in the passenger seat 😄but thanks for asking!

**Reply by Andoramb** [reply]
I'll try to debug TGW in the meantime

**Reply by Andoramb** [reply]
*Edit:
Fixed TGW in docker now 👍  (I had to completely recreate the container)
- Chat completions should work (assuming it gets the right payload and configurations)
- Listing the running models also returns the found models 

I might have some ideas for the future, but let me know how it works out for you 😀

**Reply by allanbunch** [reply]
Awesome! Thanks for the update. 😀 I'll check LM Studio too!

Future ideas would be fantastic. Thanks for this one. Cheers!
